### ISSUE TITLE
Show only windows of current workspace

### ISSUE THREAD

**Issue by 2i** [micheleg/dash-to-dock#204]
I think it would improve the usability of the extension if it had an option (in the settings) to only show applications on the current workspace in the dock.

The scenario which is inconvenient without the feature is: when you click on an icon of an application which is already running but has no window on the current workspace then you will be navigated to the the workspace which contains such a window. But if you use workspaces to work on separate things you rather would like to open a new window of the application (for example a browser window). The navigation to another workspace is distracting and you have to find your previous workspace again.

To accomplish the desired behavior without the feature you have to right click on the application icon and click on “new window” if the application is not already open on the current workspace. Otherwise you have just to left click on the icon. Currently you cannot see in the dock whether the application is already open on the current workspace, so you have to remember.

The feature would also reduce the number of icons shown in the dock and so improve lucidity.

Favorite applications in the dock instead should be shown on every workspace in the dock in my opinion.

**Comment by micheleg** [comment]
As far as I know this extension is compatible with dash to dock and does part of what you are asking https://extensions.gnome.org/extension/887/workspace-isolated-dash/.

The main problem I see with your suggestion is that there's no way to know if clicking will bring up an older window or a new one. I f you want a new window you can always use ctrl+click.

This is partialy a duplicate of #194.

**Comment by 2i** [comment]
In my opinion clicking on an icon of an application should open a new window if (and only if) there is no window on the current workspace of the same application. In the other case, an existing window of the application on the current workspace should be brought to front (like it is already if there is only one workspace).

Thanks for the link to the extension. Principally it is exactly what I was think of (except the icons changing the order sometimes, but this is another bug). But I think, bringing an option into D2D would improve it even more. ;)

**Comment by fabrizio11** [comment]
I agree with this feature request. 

[Workspace-isolated-dash](https://extensions.gnome.org/extension/887/workspace-isolated-dash/]workspace-isolated-dash) works fine **but** it keeps showing opened windows of other workspaces **if** those are windows of a favorite application. So for me isn't a nice solution at all :stuck_out_tongue_closed_eyes:

**Comment by micheleg** [comment]
Hi,

I'm open to consider this option, provided that the feature can be implemented in a clean and maintainable way. However I have no time to work on it now. I'm happy to consider a pull request and to provide guidance for the implementation if anybody wants to look into it.  The  Workspace-isolated-dash extension might be a good starting point to know which things to touch but, as noticed by @fabrizio11, more details have to be tweaked.

**Comment by franglais125** [comment]
Hi,

I managed to fully implement this on a local copy (based on Workspace-isolated-dash), but unfortunately it seems there are memory leaks (it's not terrible, but still). The dash works as if only the current workspace existed.

I'll try to work on this step by step, so that it is possible to spot bugs more easily.

Meanwhile, I implemented a first step: I added an option for "Cycle through windows in current WS".
I think it can certainly be improved, and corner cases need to be looked at. It's in a fork I made recently.

@micheleg : great extension!

Cheers

**Comment by micheleg** [comment]
Hi @franglais125,

what is your plan about this feature? Do you want to replicate and extend the Workspace-isolated-dash functionalities or do you just want to improve the extension compatibility? 

You initial commit seems ok from a very quick view.  It could be included into the extension, but I'm scared by the exponential growth of options... A single "isolated workspace" option would be more desiderable from the settings point of view. 

I'm a bit concern and surprised about the memory leaks. Do you have such code somewhere?

**Comment by nsrosenqvist** [comment]
For me workspace isolated dash isn't working, it just removes the running indicator on windows of the other workspace, and yes, they aren't favorited. So this feature would be awesome for me

**Comment by franglais125** [comment]
@micheleg  In the next few weeks I will try and upload the implementation I have from "Workspace-isolated-dash" into your extension, I'll make the code prettier first.

I totally agree with you about the options though... Too many features can render the whole thing difficult to maintain (or plainly bloated). I did it like this for now so that it was obvious what had been implemented so far. Also, for people using "Workspace-isolated-dash", the "Cycle windows in current WS" feature can prevent the shell from changing WS when a window is already open in another WS, so it can be useful on its own.

Regarding the memory leak, it only happens with the full isolation implementation. I am using gnome-shell 3.14, so maybe a few things have improved since. I don't know if it can qualify as a leak, but after a few hours of use the shell was using ~500MB of RAM.
Also, I already had this "leak" when using the "Workspace-isolated-dash" extension...
I initially started to work on this to see if the leak could be prevented by having isolation directly in your extension. I'm sure you'll be able to provide useful comments once I upload the whole thing.

I'll get gnome-shell 3.20 first, and try it there!

Cheers

**Comment by franglais125** [comment]
@micheleg I now committed the code to introduce workspace isolation in the dash. Would you mind taking a look? As it is now, it's ON by default the whole time. This is not the intended behaviour, but for now it does the job. Later on we can see how to go about including it in the settings.

[https://github.com/franglais125/dash-to-dock/commit/05972509838a01957a2133657479e1edbc9ef215](url)

Ideally the two features I added (cycle through windows in current WS and the isolation) would be a unique option, right?

**Comment by micheleg** [comment]
I haven't tested your code, but commit https://github.com/franglais125/dash-to-dock/commit/05972509838a01957a2133657479e1edbc9ef215 seems correct and clear. 

Yes, I would make a single option for both of them. 
Commit https://github.com/franglais125/dash-to-dock/commit/6f1332acf2f741e7d71105c17676185eb012a946  instead  looks a bit "dirty" at the moment, but I gess it will be improved when you merge the changes under a single option. 

From a UI perspective, I think this should be sub option of the "show running applications" switch.

The things that has always annoyed me a bit about this isolation, is the continuous resize and change of the dock whenever I switch workspace: do you find this ok, do you have any idea about how to improve the transition?

**Comment by franglais125** [comment]
Hi, so, I have quite a few comments:
1) I reverted the previous commits to pull in some commits from your master branch. This way the merging should be easier. E.g. some files were split, etc.

2) https://github.com/franglais125/dash-to-dock/commit/1f959b3e5f500abcb284043c0b5f4af03955432f https://github.com/franglais125/dash-to-dock/commit/6d3af2f463aabcc6ccc3d6054e81a19b9b842da9 I added an option under "Show running applications". Now the whole code follows this setting.

3) https://github.com/franglais125/dash-to-dock/commit/405266400d259cf0807b7182283185a52995e4d1 I did again the "dirty" part. Essentially I extended the capability of "getInterestingWindows()", to take the settings as an input. I hope this is cleaner? Also, this automatically fixes other problems, for instance counter indicators now follow the # of windows in each workspace.

4) https://github.com/franglais125/dash-to-dock/commit/b8ba4d58f39c881f2a4bb432b62e708a1eadee84 Took care of Nautilus when it is used for Desktop icons.

5) So far I didn't notice the dash changing size since I always use it in Panel mode... I just tried it the other way and yes, it can get annoying. I'm not sure this has a simple solution.
One thing we could do is set the position of the top part of the dock to follow the size of the dock from the workspace with most apps. This way at least the favourite apps won't be changing position, only the bottom would extend. However, I can see this behaving badly if someone has too many opened apps (with size not being fixed, etc.).

Generally speaking I think the implementation is fairly complete as it is. If someone doesn't like it, it can be turned off and it wouldn't bother (I set it as off by default by the way). Now, I'm sure there will be corner cases to fix as more people try different configurations.

I can at least think of one thing to improve: right-clicking on an icon should display a list of windows in the current WS only (it shows all windows). However, I'd rather make sure you like the way I'm doing this, and then focus on smaller problems.

Comments? Further improvements before we can try a PR?

**Comment by franglais125** [comment]
@micheleg: I was waiting a bit to see if I got any more feedback. Seeing as there isn't anything new, I'd like to try a PR for this.

I just have a last idea regarding what you mentioned about the dash changing size. This is of course due to the number of icons being different for each workspace, depending on the number of non-favorite apps being open.
Maybe we could show the icons for all running apps (for all WS), but we grey out or reduce the opacity for the icons of apps that are not in the current WS. This would apply to non-favorite apps. This way the dash size doesn't change while switching to another workspace, and it is clear which apps are in another workspace.

Even if we don't do that, and we show apps from all workspaces, the other part of the commits I suggested is that counter indicators count windows in the current workspace, and that on its own helps a lot with isolation.

I'd like to add that this is the last PR I have planned, I'm not working on other features, so I won't be dropping commits and PRs every other day. I intend to help you maintain/fix bugs on these things.

**Comment by nsrosenqvist** [comment]
@franglais125, I'd much rather have the dash resize, it's actually what I'd expect and want it to do. If I want to isolate the dash per workspace in the first place it doesn't make sense to show the other windows on other workspaces at all, greyed out or not, since the feature is supposed to isolate the dash.

If you still want to include it as a feature then please make it an option since it's not how I'd want it and it's counter intuitive to the isolation.

**Comment by franglais125** [comment]
@nsrosenqvist Sure, I'd rather have that too, and that's how I implemented it.
However, I was waiting to get further (or final hopefully) comments from @micheleg so that the PR is accepted. Changing a few things here and there is fine, but if something big need to be reworked, I'd rather finish it properly first.

By the way, @nsrosenqvist, would you mind testing it? It is working fine for me, but maybe you find a bug in the implementation of this feature, or you have suggestions. You can find it here: https://github.com/franglais125/dash-to-dock (zip: https://github.com/franglais125/dash-to-dock/archive/master.zip).
After installing from zip, you might need to run this:
`cd ~/.local/share/gnome-shell/extensions/dash-to-dock@micxgx.gmail.com`
`glib-compile-schemas schemas/`

**Comment by nsrosenqvist** [comment]
Tried installing it on GNOME-Shell 3.16, but couldn't activate it since it's an unsupported version. Don't have time to install a vm to test, I could test it in a couple of weeks when I have had time to upgrade to Ubuntu GNOME 16.04. Sorry!

**Comment by micheleg** [comment]
Sorry for the late reply. I haven't had the chance to look at your code yet. However, the point of avoiding the dash resizing was just and excuse. Let's go for the current behaviour. I'll try to have a look and hopefully merge this code sometime this week.

Please if you have time do an extensive testing with the different combination of  settings.

**Comment by franglais125** [comment]
@micheleg No worries. I didn't mean to rush you, we are all just eager to improve your extension!

I still have some comments though, related to bugs and implementation.

1. I have tested the code quite a bit in Gnome 3.18 and 3.20. For what versions of the Shell do you intend to support this feature? If you need me to, I can help to introduce this to older versions (I have it working on 3.14, 3.16 too, though with modifications).

2. I just fixed a bug related to closing windows from context menu (70dc292107e2b3d44efb4ce07134922ffa5126b0)

3. If someone wants to test this feature, keep in mind that some apps are single-windowed (e.g. Rhythmbox). If you open Rhythmbox (and is favorite) in workspace 1, and then you move to WS 2, clicking on Rhythmbox icon won't open a new window. This behavior seems reasonable to me.
Also, be aware that Nautilus has a bug that prevents the isolation from working as intended (bug report: https://bugzilla.gnome.org/show_bug.cgi?id=766808). So don't test it with Nautilus if the version is 3.20 or 3.18 (or don't blame the isolation!).

4. The implementation I have done takes care of the _Dash_. The behavior in the _overview_ is not fully isolated. For example:

- open gedit, move to another workspace
- open the overview and type "gedit"
- hit enter to open the app
- you will be shown the previous gedit window (in the other workspace).

I'd assume that one would want to open a new window of the app, when working in an isolated workspace.
I can implement this too (just like workspace-isolated-dash), it just means extending the Shell.App.prototype.activate function. I can prepare a patch for this too, but I wanted to know what people thought before doing it. The extension is very easy, we just need to take care of managing enable() and disable() properly. @micheleg , what do you think?

**Comment by franglais125** [comment]
This patch a8fa9bdb093e78769b9b6c4ff1706a415855107a takes care of point 4 from my previous comment.
Also, it binds the settings, so that toggling the isolation option has an immediate effect.

**Comment by nsrosenqvist** [comment]
@franglais125 I'm testing it on 3.16 and I'm loving it, works exactly as I expect it would, the overview too.

**Comment by franglais125** [comment]
I guess we can now close this issue?

@micheleg Thanks a lot for reviewing and merging this feature!!

**Comment by michael-rapp** [comment]
I just tested out this new feature and seems to work like a charm. Thanks a lot for implementing this. This was one of my personal most-wanted features in GNOME at all.

**Comment by micheleg** [comment]
Closing as implemented by #325.
